asserts/snapasserts/validation_sets.go

```diff
@@ -383,194 +525,417 @@ func (v *ValidationSets) Add(valset *asserts.ValidationSet) error {
 	return nil
 }
 
-func (v *ValidationSets) addSnap(sn *asserts.ValidationSetSnap, validationSetKey string) {
-	rev := snap.R(sn.Revision)
-	if sn.Presence == asserts.PresenceInvalid {
+// TODO: maybe a method on snapContraints?
+func addComponents(sc *snapConstraints, comps map[string]asserts.ValidationSetComponent, validationSetKey string) {
+	for name, comp := range comps {
+		addComponent(sc, name, comp, validationSetKey)
+	}
+}
+
+// TODO: maybe a method on snapContraints?
+func addComponent(sc *snapConstraints, compName string, comp asserts.ValidationSetComponent, validationSetKey string) {
+	rev := snap.R(comp.Revision)
+	if comp.Presence == asserts.PresenceInvalid {
 		rev = invalidPresRevision
 	}
 
-	rc := &revConstraint{
-		validationSetKey:  validationSetKey,
-		ValidationSetSnap: *sn,
+	rc := revConstraint{
+		validationSetKey: validationSetKey,
+		presence:         comp.Presence,
 	}
 
-	cs := v.snaps[sn.SnapID]
+	compRef := naming.NewComponentRef(sc.name, compName)
+
+	cs := sc.componentConstraints[compName]
 	if cs == nil {
-		v.snaps[sn.SnapID] = &snapContraints{
-			name:     sn.Name,
-			presence: sn.Presence,
-			revisions: map[snap.Revision][]*revConstraint{
-				rev: {rc},
+		sc.componentConstraints[compName] = &componentConstraints{
+			constraints: constraints{
+				name:     compName,
+				presence: comp.Presence,
+				revisions: map[snap.Revision][]revConstraint{
+					rev: {rc},
+				},
+				compRef: &compRef,
+				snapRef: sc.snapRef,
 			},
 		}
 		return
 	}
 
 	cs.revisions[rev] = append(cs.revisions[rev], rc)
-	if cs.presence == presConflict {
-		// nothing to check anymore
-		return
-	}
+
 	// this counts really different revisions or invalid
 	ndiff := len(cs.revisions)
 	if _, ok := cs.revisions[unspecifiedRevision]; ok {
 		ndiff--
 	}
-	switch {
-	case cs.presence == asserts.PresenceOptional:
-		cs.presence = sn.Presence
-		fallthrough
-	case cs.presence == sn.Presence || sn.Presence == asserts.PresenceOptional:
-		if ndiff > 1 {
-			if cs.presence == asserts.PresenceRequired {
+
+	cs.presence = derivePresence(cs.presence, comp.Presence, ndiff)
```

Comment:
This last bit seems to not be covered by tests. Is this missing a test case?

Reply:
Added test here: ece0c1918b964cae10c9be90357f744bc8a0c1b7